src/shared/generator.c

```diff
@@ -708,6 +709,84 @@ int generator_hook_up_pcrfs(
+                const char *where,
+                const char *target,
+                const char *fstype) {
+
+        _cleanup_free_ char *where_unit = NULL, *instance = NULL;
+        int r;
+
+        assert(dir);
+        assert(where);
+
+
+        if (isempty(fstype) || streq(fstype, "auto"))
+                return log_warning_errno(SYNTHETIC_ERRNO(EOPNOTSUPP), "Couldn't determine filesystem type for %s, \nQuota cannot be ... activated", what);
+        else if (!fstype_needs_quota(fstype))
+                return log_warning_errno(SYNTHETIC_ERRNO(EOPNOTSUPP), "Quota was requested for %s, but not supported, \nignoring: %s", what, fstype);
+
+        r = unit_name_from_path(where, ".mount", &where_unit);
+        if (r < 0)
+                return log_error_errno(r, "Failed to make unit name from path '%s': %m", where);
+
+        /* quotacheck unit for system root */
+        if (path_equal(where, "/"))
+                return generator_add_symlink(dir, where_unit, "wants", SPECIAL_QUOTACHECK_ROOT_SERVICE);
+
+        r = unit_name_path_escape(where, &instance);
+        if (r < 0)
+                return log_error_errno(r, "Failed to escape path '%s': %m", where);
+
+        r = generator_add_ordering(dir, where_unit, "After", SPECIAL_QUOTACHECK_SERVICE, instance);
+        if (r < 0)
+                return r;
+
+        if (target) {
+                r = generator_add_ordering(dir, target, "After", SPECIAL_QUOTACHECK_SERVICE, instance);
+                if (r < 0)
+                        return r;
+        }
+
+        return generator_add_symlink_full(dir, where_unit, "wants", SYSTEM_DATA_UNIT_DIR "/" SPECIAL_QUOTACHECK_SERVICE, instance);
+}
+
+int generator_hook_up_quotaon(
+                const char *dir,
+                const char *where,
+                const char *target) {
+
+        _cleanup_free_ char *where_unit = NULL, *instance = NULL, *quotaon_unit_instance = NULL;
+        int r;
+
+        assert(dir);
+        assert(where);
+
+        /* quotaon unit for system root is not instantiated */
+        if (path_equal(where, "/"))
+                return 0;
+
+        r = unit_name_from_path(where, ".mount", &where_unit);
+        if (r < 0)
+                return log_error_errno(r, "Failed to make unit name from path '%s': %m", where);
+
+        r = unit_name_path_escape(where, &instance);
+        if (r < 0)
+                return log_error_errno(r, "Failed to escape path '%s': %m", where);
+
+        r = unit_name_replace_instance(SPECIAL_QUOTAON_SERVICE, instance, &quotaon_unit_instance);
+        if (r < 0)
+                return log_error_errno(r, "Failed to instantiate '%s' for '%s': %m", instance, SPECIAL_QUOTAON_SERVICE);
+
+        r = generator_add_ordering(dir, where_unit, "Before", SPECIAL_QUOTAON_SERVICE, instance);
+        if (r < 0)
+                return r;
+
+        return generator_add_ordering(dir, quotaon_unit_instance, "After", SPECIAL_QUOTACHECK_SERVICE, instance);
```

Comment:
so these ordering deps. why generate them dynamically?

you could instead insert in quotaon@.service just:

```
After=%i.mount
After=systemd-quotacheck@%i.service
```

which I think would be much nicer because more discoverable.